### ISSUE TITLE
Finetuning Keras Model

### ISSUE THREAD

**Issue by mtamburrano** [fchollet/keras#3139]
I want to finetune a Keras Model (defined with the new functional API).
I can correctly finetune a sequential model using @joelthchao's code [here](https://github.com/fchollet/keras/issues/2640#issuecomment-217346104) but I'm not sure what's the best way to finetune the new Model type.
Actually I'm loading a model defined on a json file with `model_from_json`, so the result is the model already built and I cannot figure how to build an additionally layer, remove the last one (or two, with the activation) and substitute it with the new one.

I have a workaround, I added on this [line](https://github.com/fchollet/keras/blob/master/keras/engine/topology.py#L2385) the following code:

```
if layer.output_shape[1] != weight_values[1].shape[0]:
    print ("SKIPPING LOAD ON %s" % (layer.name))
    continue
```

and I manually changed the json adjusting the shapes as I want. So when the `load_weights` is called, essentially all the weights are loaded except those of the layers with a different shape than the original one.

Is there a cleaner way to do it without the need to manually modify the json?

**Comment by EnviousElitist** [comment]
Also interested in fine tuning from configs, as I prefer serializing all the different configurations to disk rather than committing the code for each minor modification.  Perhaps, a flag / separate function to load a config to yield a graph of layers (e.g. what is fed into the functional API), or an alterative save format.  

@mtamburrano, you should link to the file version for the current commit, e.g. [like this](https://github.com/fchollet/keras/blob/27c83c693d76c84cd704ca467d29f39dbdf4c855/keras/engine/topology.py#L2385)  

Manually modifying the JSON indeed gets very painful if one is interested in more than just swapping out the number of outputs.

**Comment by arel** [comment]
@mtamburrano I've been also trying to fine-tune a Functional-API-based model, and was finding it difficult to have to load the weights on the original architecture then alter the layers.

An alternative may be to load model weights by name. Then you can keep the names for the layers you want to load, and change the names for any altered layers. This way we avoid the layer surgery. PR #3488 should allow this, in case it helps anyone trying to do the same thing.

**Comment by Patriciasr92** [comment]
Hey everybody!

I want to finetune the last FC layer (before the Softmax) of a VGGFACE model (made with functional API)  and change the Softmax layer so instead of classify 2622 classes only classify 8 (I am trying to do a Facial Expression Recognition System). 

I tried different approaches such as:

- Separate the model in two submodels, the part of convolutions and the FC part. But then I can't merge them together.
- Tried to freeze the Convs layers (Trainable=False) and train the FC layers. But i've got an error of dimensions.

This is what I have : http://stackoverflow.com/questions/40692495/keras-error-with-training-dimension-is-not-what-is-expected

If anyone can look what I have to change in the code in order to achieve that i will always be grateful.

Thank you.

**Comment by valtron** [comment]
I'm using this. I'd be nice if something similar was available in Keras.

```
# Sample usage:
# 1. Use 'layer_name' as output (layer index works too)
model2 = surgery(model, ['layer_name'])
# 2. Two outputs
model3 = surgery(model, ['layer_name', 'other_layer'])
# 3. Select an output, and add another layer (or model) to it
model4 = surgery(model, [('layer_name', Dense(32, activation = 'softmax'))])

def surgery(model, outspecs):
	def _connect_out(outspec):
		if not isinstance(outspec, tuple):
			outspec = (outspec, None)
		(name, connect) = outspec
		o = _get_layer(model, name).output
		if connect is not None:
			o = connect(o)
		return o
	
	return Model(model.inputs, list(map(_connect_out, outspecs)))

def _get_layer(model, name_or_idx):
	if isinstance(name_or_idx, int):
		return model.get_layer(index = name_or_idx)
	return model.get_layer(name_or_idx)
```
